subsys/fs: Make FAT FS formatting on mount optional

de-nordic opened the pull request:
The fs_mount, for FAT file system, would attempt to format storage device
when mounting fails with "No System" error.  This may not always be
desired and this commit allows to choose whether formatting will be
performed in such case.

~~The FS_FATFS_MOUNT_FORMATS_WHEN_NO_FS Kconfig option, that controls this
behaviour, has been defaulted to "yes" to maintain compatibility with
previous versions.~~
The `FS_FATFS_MOUNT_MKFS` Kconfig option has been added to exclude code that formats volume when expected FAT FS.
The `FS_FATFS_READ_ONLY` Kconfig option has been added to exclude functions that perform write operations from compilation.

New field  `flags` in `fs_mount_t` structure has been added that allows to provide mount flags. Two new flags have been defined:
-  `FS_MOUNT_FLAG_NO_FORMAT` that prevents `fs_mount` from formatting the volume when no system is found
- `FS_MOUNT_FLAG_READ_ONLY` that mounts file system as read-only

TODO: 
- [x]  Make FATFS formatting optional per mount point;  
- [x]  Make it possible to turn off formatting, at compile time, when it is not needed by device to save some ROM space;
- [x]  Fix error codes returned by fs_mount (partially done here #28643)
- [x]  Fix documentation (partially done here https://github.com/zephyrproject-rtos/zephyr/pull/28643)
- [x]  Tests for new FS_MOUNT_FLAG_* flags.

SUPPORTS: https://github.com/zephyrproject-rtos/zephyr/pull/28320#issuecomment-719992771

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
subsys/fs: Make FAT FS formatting code optional

The commit adds FS_FAFTFS_MKFS Kconfig option; the option, y by default,
allows to include mkfs supporting code, within fs_mount callback,
for ELM FAT file system driver.

When formatting of FAT FS volumes is not desired, this option may be
deselected to slightly reduce code size.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

pabigot reviewed (commented):
As a feature this is good, but I would rather see it as an option that's part of `struct fs_mount`.

@nvlsianpu would you accept that path instead?

nvlsianpu commented:
^^ Yes, I would accept that.

de-nordic commented:
@pabigot, @nvlsianpu  Configuring it as a  mount option is fine. What I wanted to achieve here is also wanted a prelude to making formatting optional code.
Sometimes you just want to be able to access FATFS on SD or other media and to not care for it to be formatted by the device as you do that on deployment, and the storage is not removable in field. This means that your device carries some code that will never be executed and you could shave it off or reuse it.

pabigot commented:
OK, then I would prefer that path, perhaps by adding a `flags` field do the `fs_mount` structure.  It gets a little tricky since the default (zero value) must be to do the format, but naming the option `FS_MOUNT_FLAG_NO_FORMAT` could handle that.  `FS_MOUNT_RDONLY` is another flag that could be useful.

The `fs_mount` documentation may need to be updated so failure to mount due to lack of a formatted file system can be distinguished from other failure modes.  Currently the documentation is incorrect (`-ERRNO` is not a value).

@de-nordic do you want to proceed with that path?  Making format support optional seems reasonable and this could be part of that.

nvlsianpu commented:
I think this option might be  part of `struct fs_mount`  in generral.
And then be also selected via Kconfig for auto-mount feature,

pabigot commented:
> And then be also selected via Kconfig for auto-mount feature,

Is this partly in support of the fs backend for logging?

I'm not a fan of having a global flag like "do (not) format file system on mount".  If the logging file system is going to be auto-mount and auto-format, then I'm going to want it to be a separate filesystem from the one where I store critical data that I do not want to be formatted for me.  File systems may fail to mount because of an error, but could be recovered through an API call while reformatting it would destroy any recoverable data.

de-nordic commented:
> 
> @de-nordic do you want to proceed with that path? Making format support optional seems reasonable and this could be part of that.

Yes lets do that. So the changes would be:

- Make FATFS formatting optional per mount point;
- Make it possible to turn off formatting, at compile time, when it is not needed by device to save some ROM space;
- Fix error codes returned by fs_mount
- Fix documentation.

@pabigot does the list cover everything?

pabigot commented:
> @pabigot does the list cover everything?

I think that covers everything for one PR, as long as the solution for optional formatting is extensible to other options.  Read-only mount can be done separately when it's needed.

de-nordic commented:
> 
> 
> > @pabigot does the list cover everything?
> 
> I think that covers everything for one PR, as long as the solution for optional formatting is extensible to other options. Read-only mount can be done separately when it's needed.

I think that the Read-only feature should be handled separately.

nvlsianpu commented:
> Is this partly in support of the fs backend for logging?

Shouldn't be so dedicated - this was my fastest thought - as I see discussion went in a good direction.

de-nordic committed:
subsys/fs: Make FAT FS write support optional

The commit adds FS_FATFS_READ_ONLY Kconfig option; the option, when
selected, excludes write supporting code within ELM FAT driver.

When write support to FAT FS volumes is not desired, this option may be
selected to slightly reduce code size.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
fs: Add support for mount flags

The flags field has been added to fs_mount_t structure, accompanied
with two new flags:
  FS_MOUNT_FLAG_READ_ONLY -- mount fs as read only
  FS_MOUNT_FLAG_NO_FORMAT -- do not format volume when system not found

Code supporting the flags has been added to FS layer and drivers for
LittleFS and FAT FS.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
**Some stats for Kconfig options on nrf52840dk**

smp_svr with default configuration and FAT FS enabled:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       36196 B         1 MB      3.45%
            SRAM:        9872 B       256 KB      3.77%
        IDT_LIST:          56 B         2 KB      2.73%
```

with `CONFIG_FS_FATFS_NO_FORMAT` deselected:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       34664 B         1 MB      3.31%
            SRAM:        9872 B       256 KB      3.77%
        IDT_LIST:          56 B         2 KB      2.73%
```
with `CONFIG_FS_FATFS_READ_ONLY` selected:
```
           FLASH:       32824 B         1 MB      3.13%
            SRAM:        9872 B       256 KB      3.77%
        IDT_LIST:          56 B         2 KB      2.73%
```
Builds performed without https://github.com/zephyrproject-rtos/fatfs/pull/4 which excludes supporting code from ELM FAT FS driver.

de-nordic committed:
tests/fs/fs_api: Add API tests fo FS_MOUNT_FLAG_READ_ONLY

The commit adds read-only mount scenario tests to fs_api test
suite.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/fs/littlefs: Add tests for FS_MOUNT_FLAG_*

Adds various tests for mounting LittleFS with FS_MOUNT_FLAG_NO_FORMAT
and FS_MOUNT_FLAG_READ_ONLY, and operations on read-only mounted
file system.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/fs/fatfs: Add tests for FS_MOUNT_FLAG_*

Adds various tests for mounting ELM FAT FS with FS_MOUNT_FLAG_NO_FORMAT
and FS_MOUNT_FLAG_READ_ONLY, and operations on read-only mounted
file system.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

pabigot commented on include/fs/fs.h:
```diff
@@ -71,6 +71,11 @@ enum {
+/* Do not format volume when file system not found */
+#define FS_MOUNT_FLAG_NO_FORMAT		1
+/* Mount for read-only, this incudes FS_MOUNT_FLAG_NO_FORMAT */
+#define FS_MOUNT_FLAG_READ_ONLY		3
```
This is a little opaque.  How about:
```
#define FS_MOUNT_FLAG_NO_FORMAT_MASK BIT(0)
#define FS_MOUNT_FLAG_READ_ONLY_MASK BIT(1)

#define FS_MOUNT_FLAG_NO_FORMAT FS_MOUNT_FLAG_NO_FORMAT_MASK
#define FS_MOUNT_FLAG_READ_ONLY (FS_MOUNT_FLAG_READ_ONLY_MASK \
                                 | FS_MOUNT_FLAG_NO_FORMAT_MASK)
```
separating the specific bits from the user-visible mount options.

Or just use the two base flags and document that read-only implies no-format.

de-nordic replied:
Will use.

pabigot commented on subsys/fs/fs.c:
```diff
@@ -148,6 +148,10 @@ int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
 	}
 
 	zfp->mp = mp;
+	if (mp->flags & FS_MOUNT_FLAG_READ_ONLY &&
```
A reason combining multiple bits into the named flag is bad: this test will pass if `FS_MOUNT_FLAG_NO_FORMAT` is the only mount flag, which would be wrong.

pabigot replied:
It's also an argument for what MISRA would require:

    ((mp->flags & FS_MOUNT_FLAG_READ_ONLY) == FS_MOUNT_FLAG_READ_ONLY)

would work where

    ((mp->flags & FS_MOUNT_FLAG_READ_ONLY) != 0)

would not.

de-nordic replied:
Bummer. Will fix.

pabigot reviewed (changes requested):
Using multiple bits in a named mask intended to identify only a single condition is a problem.

pabigot commented on subsys/fs/fs.c:
```diff
@@ -148,6 +148,10 @@ int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
 	}
 
 	zfp->mp = mp;
+	if (((mp->flags & FS_MOUNT_FLAG_READ_ONLY) == FS_MOUNT_FLAG_READ_ONLY) &&
```
I really think this should be `(mp->flags & FS_MOUNT_FLAG_READ_ONLY_MASK) != 0)` or something else that only tests the read-only bit.  This part of the code doesn't care whether the file system was flagged with no-format.  That the two flags are combined are only a convenience for the user, one that I'm not convinced is justified.

It's reasonable IMO to keep the flags completely separate, and simply error if somebody tries to mount a file system read-only in a situation where it would have to be initialized.  Here having the wrapper implicitly add the no-format flag when read-only seems less confusing.

pabigot commented on subsys/fs/fat_fs.c:
```diff
@@ -119,6 +119,7 @@ static int fatfs_close(struct fs_file_t *zfp)
 	return translate_error(res);
 }
 
+#if !defined(CONFIG_FS_FATFS_READ_ONLY)
```
This looks related to the other proposal; I think it'd be totally reasonable to implement this as:
```
static int fatfs_unlink(struct fs_mount_t *mountp, const char *path)
￼{
    FRESULT res = -ENOTSUP;
#if !defined(CONFIG_FS_FATFS_READ_ONLY)
    // main body
    res = translate_error(res);
#endif
    return res;
}
```

which eliminates the need to make any changes at all to the `fs_file_system_t` objects.

de-nordic replied:
> 
> 
> This looks related to the other proposal; I think it'd be totally reasonable to implement this as:
> 
> ```
> static int fatfs_unlink(struct fs_mount_t *mountp, const char *path)
> ￼{
>     FRESULT res = -ENOTSUP;
> #if !defined(CONFIG_FS_FATFS_READ_ONLY)
>     // main body
>     res = translate_error(res);
> #endif
>     return res;
> }
> ```
> 
> which eliminates the need to make any changes at all to the `fs_file_system_t` objects.

I can do that, no problem for me.

The #29396  has not been created because of this. Basically I wanted to get rid of the NULL check within fs_ functions, and I have offered something other in exchange.
You may not agree with me but I think the NULL checks are not needed there as I believe that developers exactly know which functions are offered by the system driver or at least will gain the knowledge quickly.
Basically: when writing embedded software you get -ENOTSUP, specifically in the meaning "not implemented". once  then you have to reorganize to solve the problem and you newer see it again.
If you green light me in #29396 to drop defaulting not implemented functions, while being allowed to remove/make optional the NULL checks: that will be exactly what I will do.

pabigot reviewed (approved):
I don't see anything that merits blocking this, so approved.

I do think it'd be easier to maintain by making the global read-only support only eliminate the code within the functions, rather than remove the functions entirely.  Then it becomes less clear that #29396 is worthwhile (I still don't really like it, though again not enough to reject it outright).

pabigot reviewed (changes requested):
Documentation update needed, maybe config option changes.

pabigot commented:
@nvlsianpu @nashif I've dismissed your reviews because this PR has changed significantly in the 5-6 weeks since your approvals.  It's still in flux, but I think the enhancements improve it.  Please take another look.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -14,6 +14,20 @@ if FAT_FILESYSTEM_ELM
+	help
+	  Enables mkfs support within ELM FAT FS.
+
+config FS_FATFS_MOUNT_MKFS
```
It is on the FAT file system level because it is often used on external devices that you may not want to be formatted by your device when connected. It allows stripping a few kib, from ROM, of code you are never going to use...

de-nordic replied:
OK, the comment on size reduction has been missing...

pabigot replied:
What external devices?  Is this for somebody plugging a USB thumb drive into a device running Zephyr?

If there's a good reason, ok, though I still think FS-wide would be more natural (what if Zephyr's configured for a littlefs file system? then the external device will still be formatted).  Also the impact of this on the mount flags does need to be made clear, probably both at the Kconfig and at the flags.

de-nordic replied:
> 
> 
> What external devices? Is this for somebody plugging a USB thumb drive into a device running Zephyr?
> 
Or SD.

>If there's a good reason, ok, though I still think FS-wide would be more natural (what if Zephyr's configured for a >littlefs file system?

Yes, but what if you have a device that internally uses LittleFS for configuration and calibration, etc. and will format internal partition when placed in burn-in environment. In such case you may need to have mkfs capability in LittleFS, while does not desire such within FAT FS or other systems.

pabigot replied:
If you don't want mkfs for FatFs you disable `FS_FATFS_MKFS`.  If you have mkfs support in FatFs, and you don't want a particular mount to autoformat, you turn it off in the mount options, regardless of file system type.

I'm still not seeing why FatFs alone should have `FS_FATFS_MOUNT_MKFS` to disable auto-format completely when mkfs is available, regardless of mount flags that may request it.

Can you provide a more detailed use case for this?

de-nordic replied:
> 
> 
> If you don't want mkfs for FatFs you disable `FS_FATFS_MKFS`. If you have mkfs support in FatFs, and you don't want a particular mount to autoformat, you turn it off in the mount options, regardless of file system type.
> 
I can reduce code size bu rejecting code I will never use.

> I'm still not seeing why FatFs alone should have `FS_FATFS_MOUNT_MKFS` to disable auto-format completely when mkfs is available, regardless of mount flags that may request it.
> 
OK, so I have not predicted usage of LittleFS on external device, which is my oversight, and I find no other way to initialize  LittleFS on soc, and I need it.

> Can you provide a more detailed use case for this?
You are building device that displays simple led matrix animations from SD card and information like birth day dates, current time, some other calendar events, it stores data you want to be persistent, even if SD is removed, on internal littlefs formatted partition; it uses SD card in read-only mode for getting animations; formatting the SD is useless, e.g. when system is broken, as it will contain no images afterwards anyway.

pabigot replied:
If the device never formats the SD card, why not just turn off FAT mkfs support?  If the device needs FAT mkfs support for another filesystem, why not just set the mount flags for the SD card to inhibit formatting that mount?  Why do you need to have a special flag that says "mkfs is supported, but don't allow it to be used in mounts even if the mount specifies it's ok"?

I must still be missing something....

nvlsianpu replied:
I think prompt there should be "**Allow** to format unformatted volume when mounting fails" or **Support**

`FS_FATFS_MKFS` is switch for enabling MKFS procedures in fatfs - Should be considered that this option might depend on availability of MKFS procedures(`depend on FS_FATFS_MKFS`), if volume formatting without  mounting FS immediately can e a user-case.

pabigot replied:
OK, I get it now: the first one is the hidden one that forwards to FatFs; this one is the public one that would be set to `n` if formatting on mount is never desired.

nvlsianpu commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -14,6 +14,20 @@ if FAT_FILESYSTEM_ELM
 menu "ELM FAT file system settings"
 	visible if FAT_FILESYSTEM_ELM
 
+config FS_FATFS_MKFS
```
Where I can find translation to the value under `_USE_MKFS` flag?

de-nordic replied:
It has been exposed with this commit https://github.com/zephyrproject-rtos/fatfs/commit/320e2c06efd2eb10e5095da5d856d04e6a4e6395

pabigot commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -14,6 +14,15 @@ if FAT_FILESYSTEM_ELM
+config FS_FATFS_READ_ONLY
+	bool "Read-only support for all volumes"
+	help
+	  The option excludes write code from ELM FAT file system driver;
```
The `FS_FATFS_MKFS` option says what it translates to in FatFs.  I think this one should too:

> This option translates to _FS_READONLY within the Zephyr fork of the ELM FAT file system. The option excludes write ....

pabigot reviewed (changes requested):
One change to the Kconfig documentation, please.

I think this needs a west.yml update to pull in the FatFs enhancements.

Also the DNM annotation in the title should be removed.

I don't think the new error codes nor the new feature invoke the stable API change process, but I do think this should be mentioned in the API meeting.  We need more reviews.

de-nordic commented:
How can I weasel out of tthe documentation errors? The double-declaration does not come from this PR.

de-nordic commented:
> I think this needs a west.yml update to pull in the FatFs enhancements.
> 
Yes, but it will compile/work without it.
 
> I don't think the new error codes nor the new feature invoke the stable API change process, but I do think this should be mentioned in the API meeting. We need more reviews.

Should I add dev-review tag?

pabigot commented:
The documentation errors are fixed upstream; rebase on latest main branch when you fix the Kconfig.

I don't think this needs dev-review.

Agreed it works without the west update as long as the Kconfigs aren't changed from their default.  I guess it doesn't *have* to have the west update, if there's some reason to not do that.

de-nordic committed:
west.yml: Update modules/fs/fatfs to latest

The FAT FS driver module has been updated to latest

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
> 
> The documentation errors are fixed upstream; rebase on latest main branch when you fix the Kconfig.
> 
I keep on re-basing since morning and I keep getting the error...

> 
> Agreed it works without the west update as long as the Kconfigs aren't changed from their default. I guess it doesn't _have_ to have the west update, if there's some reason to not do that.

It should also work when enabled as it only tuns off few functions from compilation in FAT FS ELM driver, which will not be linked anyway. Nevertheless, I have updated the west.yml.

pabigot commented:
Documentation is finally building, so it's just [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/27910#discussion_r516732008) left.

pabigot commented:
@nashif @nvlsianpu please revisit, this changed a lot since your last review.

nvlsianpu commented on include/fs/fs.h:
```diff
@@ -173,6 +180,8 @@ struct fs_statvfs {
  *
  * @retval 0 on success;
  * @retval -EINVAL when a bad file name is given;
+ * @retval -EROFS if file is read-only, or when file system has been mouted with
```
this is all about when file was opened with the "write" flag

pabigot replied:
s/mouted/mounted/g;

de-nordic replied:
> this is all about when file was opened with the "write" flag

Unfortunately I do not understand.

nvlsianpu replied:
clarified off-channel

pabigot replied:
I didn't understand either, but from the patch now I do.  +1.

nvlsianpu commented:
@carlescufi Can be merged.
